File: 337_House_Robber_III.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rob(self, root):
        """
        :type root: TreeNode
        :rtype: int
            """
        if not root:
            return 0
        if (not root.left )and (not root.right):
            return root.val
        if root.left:
            left=self.rob(root.left.left)+self.rob(root.left.right)
        else: left=0
        if root.right:
            right=self.rob(root.right.left)+self.rob(root.right.right)
        else: right=0
        if (root.val+left+right)>(self.rob(root.left)+self.rob(root.right)):
            return root.val+left+right
        else:
            return self.rob(root.left)+self.rob(root.right)

File: 337_House_Robber_III.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int rob(TreeNode root) {
        Solution s=new Solution();
        int left;
        int right;
        if (root==null) return 0;
        if ((root.left==null) && (root.right==null) )
            return root.val;
        if (root.left!=null)
            left=s.rob(root.left.left)+s.rob(root.left.right);
        else left=0;
        if (root.right!=null)
            right=s.rob(root.right.left)+s.rob(root.right.right);
        else right=0;
        if ((root.val+left+right)>(s.rob(root.left)+s.rob(root.right)))
            return root.val+left+right;
        else 
            return s.rob(root.left)+s.rob(root.right);
    }
}
